File: src/python/take_home_package/take_home/base/base_queue.py
''' Copyright (c) Docugami, Inc. All rights reserved. '''

from abc import ABC, abstractmethod

class BaseQueue(ABC):
    ''' Abstract base class for queue implementations. '''
    def __init__(self, max_in_memory: int):
        self._max_in_memory = max_in_memory

    @property
    def max_in_memory(self) -> int:
        '''
            The maximum number of items that can be held in memory at any given time.
            All other items in the queue must be written to disk.
        '''
        return self._max_in_memory

    @property
    @abstractmethod
    def count(self) -> int:
        '''
            The total number of items in the queue.
        '''

    @property
    @abstractmethod
    def in_memory_count(self) -> int:
        '''
            The number of items currently in memory.
        '''

    @property
    @abstractmethod
    def on_disk_count(self) -> int:
        '''
            The number of items currently on disk.
        '''

    @abstractmethod
    def enqueue(self, value):
        '''
            Adds an item to the queue.
        '''

    @abstractmethod
    def dequeue(self) -> object:
        '''
            Removes an item from the queue and returns it.
        '''

    @abstractmethod
    def peek(self) -> object:
        '''
            Returns the first item in the queue.
        '''

File: src/csharp/Base/BaseQueue.cs
// Copyright (c) Docugami, Inc. All rights reserved.

namespace TakeHome.Base
{
    public abstract class BaseQueue<T> : IQueue<T>
    {
        protected BaseQueue(int maxInMemory)
        {
            this.MaxInMemory = maxInMemory;
        }

        public abstract int Count { get; }

        public abstract int InMemoryCount { get; }

        public abstract int OnDiskCount { get; }

        public int MaxInMemory { get; }

        public abstract T Dequeue();

        public abstract void Enqueue(T item);

        public abstract T Peek();
    }
}
